### PR TITLE
Problem: mr.developer broken by CLI arg to avoid endless loop

### PR DESCRIPTION
**Pull request by gotcha**
Solution: use environment variable instead

See #515

### CODE REVIEW

File: src/zc/buildout/buildout.py [hunk 1]
```diff
@@ -1172,11 +1172,11 @@ def _maybe_upgrade(self):
 
         # Restart
         args = sys.argv[:]
-        args.insert(1, '--restart-after-upgrade')
         if not __debug__:
             args.insert(0, '-O')
         args.insert(0, sys.executable)
-        sys.exit(subprocess.call(args))
+        env = {'BUILDOUT_RESTART_AFTER_UPGRADE': '1'}
```
Comment (perrinjerome) [inline src/zc/buildout/buildout.py:1178]: Shouldn't this be something like 
```python
        env = dict(os.environ, BUILDOUT_RESTART_AFTER_UPGRADE='1')
```
to also preserve environment from current process variables ?

File: src/zc/buildout/buildout.py [hunk 2]
```diff
@@ -2161,8 +2161,6 @@ def main(args=None):
                     _help()
                 elif orig_op == '--version':
                     _version()
-                elif orig_op == '--restart-after-upgrade':
-                    options.append(('buildout', 'restart-after-upgrade', "true"))
```
Comment (mgedmin) [inline src/zc/buildout/buildout.py:2165]: I'm curious as to what happens if a zc.buildout before this patch upgrades itself to a version after this patch.  Will the old buildout add --restart-after-upgrade to the command line, causing the new buildout to fail with an unknown command line argument error?

**Comment by mgedmin** [comment]
Do you have a link to a mr.developer bug, or a traceback, or something for us to see how mr. developer fails?

**Comment by mgedmin** [comment]
The CI failures I looked at seem to be caused by Python version deprecation warnings.  Didn't we have a fix for that in some other PR?  Perhaps, since the fix consisted of putting an environment variable asking pip not to print those warnings, the fix may have broken for the reason @perrinjerome noticed: that environment variable is now lost when you replace the entire environment.

**Comment by mgedmin** [comment]
> Didn't we have a fix for that in some other PR? 

Ah, not yet: #534.  Still, the CI fix will need the environment to be preserved.

File: src/zc/buildout/buildout.py (hunk 1 above)
Comment (gotcha) [reply src/zc/buildout/buildout.py:1178]: @perrinjerome Right, I missed that.

File: src/zc/buildout/buildout.py (hunk 2 above)
Comment (gotcha) [reply src/zc/buildout/buildout.py:2165]: @mgedmin  ``--restart-after-upgrade`` only exist in 3.0.0a1 and 3.0.0a2

I feel we do not need to cover those alpha versions. What is your take on that ?

**Comment by gotcha** [comment]
I added the link to #515 in the description.

File: src/zc/buildout/buildout.py (hunk 2 above)
Comment (mgedmin) [reply src/zc/buildout/buildout.py:2165]: Good point!  I forgot when it was introduced; seemed like ages ago.

**Comment by icemac** [comment]
Is there still something needed to be done before merging this PR?

**Comment by ZachArani** [comment]
> 
> 
> Is there still something needed to be done before merging this PR?

I second this. Just found this issue rearing its ugly head in my workflows.

**Pull request merged**